### PR TITLE
Add ability to retry nodes after errors

### PR DESCRIPTION
**Pull request by simster7**
Fixes https://github.com/argoproj/argo/issues/1603 and implements ability to retry nodes that result in error. This can be controlled with a new field:
```
  - name: whalesay
    retryStrategy:
      limit: 10
      retryPolicy: "Always"
```

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1162,11 +1182,13 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 		}
 		processedRetryParentNode, err := woc.processNodeRetries(retryParentNode, *processedTmpl.RetryStrategy)
 		if err != nil {
+			woc.log.Errorf("Error: %+v", err)
```
Comment (dtaniwaki) [inline workflow/controller/operator.go]: I moved error log to `markNodeError`.
https://github.com/argoproj/argo/pull/1694/files#diff-fcb04129f1d32e69cca32631c6586587R1426

Comment (simster7) [reply workflow/controller/operator.go]: This is a result of the merge, will remove these.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -717,10 +717,22 @@ type WorkflowStatus struct {
 	Outputs *Outputs `json:"outputs,omitempty"`
 }
 
+type RetryPolicy string
+
+const (
+	RetryPolicyAlways    RetryPolicy = "Always"
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: Simplify the enum
	RetryPolicyAlways    RetryPolicy = "Always" /
	RetryPolicyOnFailure RetryPolicy = "OnFailure"
	RetryPolicyOnError   RetryPolicy = "OnError"

**Comment by simster7** [comment]
@sarabala1979 Addressed your comment and included example. Should be ready for another review.

**Comment by dtaniwaki** [comment]
#1732 is related to the retry behavior.

**Comment by sarabala1979** [comment]
Please add some examples

**Comment by simster7** [comment]
This already has an example: https://github.com/argoproj/argo/pull/1696/files#diff-23bb5d88f557d611d6be3f6e75d825b0

**Comment by simster7** [comment]
Done

**Comment by zhujiangyou** [comment]
What is the difference between OnFailure and OnError？

**Comment by simster7** [comment]
> What is the difference between OnFailure and OnError？

On failure indicates an error within the container itself (i.e. an exit code != 0). On error indicates an error with the scheduling or execution of the container as a whole (e.g. error scheduling Pod, error pulling images, error processing artifacts, etc.)

**Pull request merged**